File: src/java_mag/Beispiel_4.java
package java_mag;

public class Beispiel_4 {

	public final String text;

	protected Beispiel_4(String text) {
		this.text = text;
	}

	public String toString() {
		return "JAVA-toString():" + text;
	}

	public String foo(String s) {
		return "JAVA-foo(String):" + text + ":" + s;
	}

	public String foo(int i) {
		return "JAVA-foo(Integer):" + text + ":" + i;
	}

	public String foo(String s, boolean b) {
		return "JAVA-foo(String, boolean):" + text + ":" + s + "/" + b;
	}

}

File: clj/beispiel_4.clj
(import java_mag.Beispiel_4)
(defn make-bsp-4 [p]
  (proxy [Beispiel_4] [(.toUpperCase p)]
    (toString [] 
      (format "CLJ-toString[]: '%s'\n  super.toString(): '%s'" 
              p
              (proxy-super toString)))
    (foo 
      ([x] (format "CLJ-foo[x]: %s:'%s'\n  super.foo: '%s'" 
                   (condp apply [x]
                     string? "STRING"
                     integer? "INTEGER") 
                   x
                   (proxy-super foo x)))
      ([s b] (format "CLJ-foo[s b]: '%s' '%s'\n  super.text: '%s'  this: '%s'" 
                     s b (proxy-super text) this)))))
(let [x (make-bsp-4 "foobar")]
  (println (str x))
  (println (.foo x "zweiundvierzig"))
  (println (.foo x 42))
  (println (.foo x "bar" true)))
